src/cli/onefuzz/job_templates/main.py

```diff
@@ -0,0 +1,277 @@
+        if container_names is None:
+            container_names = {}
+        for container_type in config.containers:
+            if container_type.name in container_names:
+                container_name = args["container_names"][container_type]
+            else:
+                container_name = _build_container_name(
+                    self.onefuzz,
+                    container_type,
+                    args["project"],
+                    args["name"],
+                    args["build"],
+                    platform,
+                )
+            containers.append(TaskContainers(name=container_name, type=container_type))
+        return containers
+
+    def _convert_args(
+        self, config: JobTemplateConfig, args: Any, platform: OS
+    ) -> JobTemplateRequest:
+        """ convert arguments from argparse into a JobTemplateRequest """
+
+        user_fields = {}
+        for field in config.user_fields:
+            if field.name not in args and field.required:
+                raise Exception("missing field: %s" % field.name)
+            value = args[field.name]
+            if value is not None:
+                user_fields[field.name] = value
+
+        containers = self._convert_container_args(config, args, platform)
+        for container in containers:
+            directory_arg = container_type_name(container.type)
+
+            self.onefuzz.containers.create(
+                container.name, metadata={"container_type": container.type.name}
+            )
+
+            if directory_arg in args and args[directory_arg] is not None:
+                self.onefuzz.containers.files.upload_dir(
+                    container.name, args[directory_arg]
+                )
+
+        request = JobTemplateRequest(
+            name=config.name, user_fields=user_fields, containers=containers
+        )
+        return request
+
+    def _process_containers(self, request: JobTemplateRequest, args: Any) -> None:
+        """ Create containers based on the argparse args """
+
+        for container in request.containers:
+            directory_arg = container_type_name(container.type)
+            self.onefuzz.logger.info("creating container: %s", container.name)
+            self.onefuzz.containers.create(
+                container.name, metadata={"container_type": container.type.name}
+            )
+
+            if directory_arg in args and args[directory_arg] is not None:
+                self.onefuzz.logger.info(
+                    "uploading %s to %s", args[directory_arg], container.name
+                )
+                self.onefuzz.containers.files.upload_dir(
+                    container.name, args[directory_arg]
+                )
+            elif container.type == ContainerType.setup and "target_exe" in args:
+                # This is isn't "declarative", but models our existing paths for
+                # templates.
+
+                target_exe = args["target_exe"]
```

Comment:
I'm not following why we're doing the special-casing in this way. Will revisit.

Reply:
In order to specify uploading a single binary, rather than a directory, for the setup container.

Reply:
Thanks. My eyes were crossing by the time I got here in the review 😆

Reply:
Okay, so handling the `target_exe`-only case makes sense.

In that case, am I correct in reading the flow as:
- If `--setup_dir` is set (and we need to upload it), then it will be be bound to `directory_arg` in this iteration of the `for` loop, and we'll take the `if` branch, wherein we'll upload the entire `setup_dir` remotely
- Else, `--setup_dir` is not set, but we still need to check for `target_exe`, and upload it (now) to the remote setup container that we created at the start of the loop iteration.